engines/vcruise/textparser.cpp

```diff
@@ -0,0 +1,322 @@
+			continue;
+
+		outC = c;
+		outState = firstCharState;
+		return true;
+	}
+
+	return false;
+}
+
+bool TextParser::isDelimiter(char c) {
+	if (c == ',' || c == '=' || c == '[' || c == ']')
+		return true;
+
+	return false;
+}
+
+bool TextParser::isWhitespace(char c) {
+	return (c == ' ') || ((c & 0xe0) == 0);
+}
+
+void TextParser::requeue(const char *chars, uint numChars, const TextParserState &state) {
+	_state = state;
+	assert(_returnBufferPos >= numChars);
+	_returnBufferPos -= numChars;
+	memcpy(_returnedBuffer + _returnBufferPos, chars, numChars);
+}
+
+void TextParser::requeue(const Common::String &str, const TextParserState &state) {
+	requeue(str.c_str(), str.size(), state);
+}
+
+void TextParser::expectToken(Common::String &outToken, const Common::String &blamePath) {
+	TextParserState state;
+	expectTokenInternal(outToken, blamePath, state);
+}
+
+void TextParser::expectShort(int16 &outInt, const Common::String &blamePath) {
+	int i;
+	expectInt(i, blamePath);
+	outInt = static_cast<int16>(i);
+}
+
+void TextParser::expectInt(int &outInt, const Common::String &blamePath) {
+	Common::String token;
+	TextParserState state;
+	expectTokenInternal(token, blamePath, state);
+
+	int result = 0;
+	bool isNegative = false;
+	uint startIndex = 0;
+	if (token[0] == '-') {
+		if (token.size() == 1)
+			error("Parsing error in '%s' at line %i col %i: Signed integer was malformed", blamePath.c_str(), static_cast<int>(state._lineNum), static_cast<int>(state._col));
+
+		isNegative = true;
+		startIndex = 1;
+	}
+
+	int base = 10;
+	bool isHex = (token.size() >= (startIndex) + 3 && token[startIndex] == '0' && token[startIndex + 1] == 'x');
+	if (isHex) {
+		startIndex += 2;
+		base = 16;
+	}
+
+	for (uint i = startIndex; i < token.size(); i++) {
+		char c = token[i];
+		int digit = 0;
+		if (c >= '0' && c <= '9')
+			digit = (c - '0');
+		else if (isHex && (c >= 'a' && c <= 'f'))
+			digit = (c - 'a') + 0xa;
+		else if (isHex && (c >= 'A' && c <= 'F'))
+			digit = (c - 'A') + 0xa;
+		else
+			error("Parsing error in '%s' at line %i col %i: Integer contained non-digits", blamePath.c_str(), static_cast<int>(state._lineNum), static_cast<int>(state._col));
+
+		if (isNegative)
+			digit = -digit;
+
+		result = result * base + digit;
+	}
+
+	outInt = result;
+}
+
+void TextParser::expectUInt(uint &outUInt, const Common::String &blamePath) {
+	Common::String token;
+	TextParserState state;
+	expectTokenInternal(token, blamePath, state);
+
+	uint result = 0;
+
+	for (uint i = 0; i < token.size(); i++) {
+		char c = token[i];
+		if (c < '0' || c > '9')
+			error("Parsing error in '%s' at line %i col %i: Integer contained non-digits", blamePath.c_str(), static_cast<int>(state._lineNum), static_cast<int>(state._col));
+
+		uint additional = c - '0';
+
+		result = result * 10 + additional;
+	}
+
+	outUInt = result;
+}
+
+void TextParser::expectLine(Common::String &outToken, const Common::String &blamePath, bool continueToNextLine) {
```

Comment:
consider `Common::String::readLine()`

Reply:
I think this is preferable because it's easier to keep up line/col error diagnostics synced up.